File: app/src/main/java/com/example/earthwallet/ui/pages/staking/StakeEarthFragment.java
package com.example.earthwallet.ui.pages.staking;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.viewpager2.widget.ViewPager2;

import com.example.earthwallet.R;
import com.example.earthwallet.Constants;
import com.example.earthwallet.bridge.services.SecretQueryService;
import com.example.earthwallet.wallet.services.SecureWalletManager;
import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;

import org.json.JSONObject;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Main staking fragment that manages ERTH token staking
 * Features tabs for Info & Rewards, Stake/Unstake, and Unbonding
 */
public class StakeEarthFragment extends Fragment {
    
    private static final String TAG = "StakeEarthFragment";
    
    // UI Components
    private TabLayout tabLayout;
    private ViewPager2 viewPager;
    
    // Data and Services
    private SecretQueryService queryService;
    private ExecutorService executorService;
    
    // Adapter
    private StakingTabsAdapter stakingAdapter;
    
    // Interface for communication with parent
    public interface StakeEarthListener {
        String getCurrentWalletAddress();
        void onStakingOperationComplete();
    }
    
    private StakeEarthListener listener;
    
    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (getParentFragment() instanceof StakeEarthListener) {
            listener = (StakeEarthListener) getParentFragment();
        } else if (context instanceof StakeEarthListener) {
            listener = (StakeEarthListener) context;
        } else {
            Log.w(TAG, "Parent does not implement StakeEarthListener");
        }
    }
    
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_stake_earth, container, false);
    }
    
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        
        // Initialize services
        queryService = new SecretQueryService(getContext());
        executorService = Executors.newCachedThreadPool();
        
        initializeViews(view);
        setupTabs();
    }
    
    private void initializeViews(View view) {
        tabLayout = view.findViewById(R.id.tab_layout);
        viewPager = view.findViewById(R.id.view_pager);
    }
    
    private void setupTabs() {
        if (tabLayout == null || viewPager == null) return;
        
        // Create adapter
        stakingAdapter = new StakingTabsAdapter(this);
        viewPager.setAdapter(stakingAdapter);
        
        // Connect TabLayout with ViewPager2
        new TabLayoutMediator(tabLayout, viewPager,
                (tab, position) -> {
                    switch (position) {
                        case 0: tab.setText("Info & Rewards"); break;
                        case 1: tab.setText("Stake/Unstake"); break;
                        case 2: tab.setText("Unbonding"); break;
                    }
                }).attach();
    }
    
    
    /**
     * Public method to refresh staking data across all fragments
     */
    public void refreshStakingData() {
        Log.d(TAG, "Refreshing staking data for all fragments");
        
        // With ViewPager2, we need to notify fragments differently
        if (stakingAdapter != null) {
            // The fragments are managed by the adapter, so we can trigger refresh
            // through the current fragment if needed
            stakingAdapter.notifyDataSetChanged();
        }
    }
    
    /**
     * Query user staking info from contract
     */
    public void queryUserStakingInfo(UserStakingCallback callback) {
        executorService.execute(() -> {
            try {
                String userAddress = SecureWalletManager.getWalletAddress(getContext());
                if (userAddress == null || userAddress.isEmpty()) {
                    Log.w(TAG, "No user address available");
                    return;
                }
                
                // Create query message: { get_user_info: { address: "secret1..." } }
                JSONObject queryMsg = new JSONObject();
                JSONObject getUserInfo = new JSONObject();
                getUserInfo.put("address", userAddress);
                queryMsg.put("get_user_info", getUserInfo);
                
                Log.d(TAG, "Querying staking contract: " + Constants.STAKING_CONTRACT);
                Log.d(TAG, "Query message: " + queryMsg.toString());
                
                JSONObject result = queryService.queryContract(
                    Constants.STAKING_CONTRACT,
                    Constants.STAKING_HASH,
                    queryMsg
                );
                
                Log.d(TAG, "Staking query result: " + result.toString());
                
                if (callback != null && getActivity() != null) {
                    getActivity().runOnUiThread(() -> callback.onStakingDataReceived(result));
                }
                
            } catch (Exception e) {
                Log.e(TAG, "Error querying staking info", e);
                if (callback != null && getActivity() != null) {
                    getActivity().runOnUiThread(() -> callback.onError(e.getMessage()));
                }
            }
        });
    }
    
    @Override
    public void onDetach() {
        super.onDetach();
        listener = null;
    }
    
    @Override
    public void onDestroy() {
        super.onDestroy();
        if (executorService != null && !executorService.isShutdown()) {
            executorService.shutdown();
        }
    }
    
    /**
     * Callback interface for staking data queries
     */
    public interface UserStakingCallback {
        void onStakingDataReceived(JSONObject data);
        void onError(String error);
    }
}

File: app/src/main/java/com/example/earthwallet/ui/pages/staking/StakeEarthFragment.kt
package network.erth.wallet.ui.pages.staking

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import network.erth.wallet.R
import network.erth.wallet.chain.Staking
import network.erth.wallet.wallet.services.ErthPriceService
import network.erth.wallet.wallet.services.SecureWalletManager
import java.math.BigInteger

/**
 * Native x/staking home: an info panel (staked / total bonded / APR / earnings)
 * above 5 tabs (Rewards, Stake, Withdraw, Redelegate, Unbonding).
 *
 * ERTH staking is native delegation; inflation of 1 ERTH/sec is distributed to
 * stakers, so APR = annual inflation / total bonded.
 */
class StakeEarthFragment : Fragment() {

    companion object {
        private const val TAG = "StakeEarthFragment"
        private const val ANNUAL_INFLATION_ERTH = 31_536_000.0 // 1 ERTH/sec * seconds/year
    }

    // Info section
    private var infoSection: LinearLayout? = null
    private var aprText: TextView? = null
    private var stakedAmountText: TextView? = null
    private var stakedAmountUsd: TextView? = null
    private var totalStakedText: TextView? = null
    private var totalStakedUsd: TextView? = null
    private var poolShareText: TextView? = null
    private var dailyEarningsText: TextView? = null
    private var dailyEarningsUsd: TextView? = null

    // Tabs
    private var tabLayout: TabLayout? = null
    private var viewPager: ViewPager2? = null
    private var stakingAdapter: StakingTabsAdapter? = null

    private var transactionSuccessReceiver: BroadcastReceiver? = null

    // Data
    private var stakedBalance = 0.0
    private var totalStakedBalance = 0.0
    private var apr = 0.0
    private var erthPrice: Double? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_stake_earth, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initializeViews(view)
        setupBroadcastReceiver()
        registerBroadcastReceiver()
        setupTabs()
        refreshStakingData()
    }

    private fun initializeViews(view: View) {
        infoSection = view.findViewById(R.id.info_section)
        aprText = view.findViewById(R.id.apr_text)
        stakedAmountText = view.findViewById(R.id.staked_amount_text)
        stakedAmountUsd = view.findViewById(R.id.staked_amount_usd)
        totalStakedText = view.findViewById(R.id.total_staked_text)
        totalStakedUsd = view.findViewById(R.id.total_staked_usd)
        poolShareText = view.findViewById(R.id.pool_share_text)
        dailyEarningsText = view.findViewById(R.id.daily_earnings_text)
        dailyEarningsUsd = view.findViewById(R.id.daily_earnings_usd)

        tabLayout = view.findViewById(R.id.tab_layout)
        viewPager = view.findViewById(R.id.view_pager)
    }

    private fun setupBroadcastReceiver() {
        transactionSuccessReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                refreshStakingData()
                Handler(Looper.getMainLooper()).postDelayed({ refreshStakingData() }, 500)
            }
        }
    }

    private fun registerBroadcastReceiver() {
        if (activity != null && transactionSuccessReceiver != null) {
            val filter = IntentFilter("network.erth.wallet.TRANSACTION_SUCCESS")
            try {
                if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.TIRAMISU) {
                    requireActivity().applicationContext.registerReceiver(transactionSuccessReceiver, filter, Context.RECEIVER_NOT_EXPORTED)
                } else {
                    requireActivity().applicationContext.registerReceiver(transactionSuccessReceiver, filter)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to register broadcast receiver", e)
            }
        }
    }

    private fun setupTabs() {
        val tabLayout = this.tabLayout ?: return
        val viewPager = this.viewPager ?: return

        stakingAdapter = StakingTabsAdapter(this)
        viewPager.adapter = stakingAdapter

        TabLayoutMediator(tabLayout, viewPager) { tab, position ->
            tab.text = when (position) {
                StakingTabsAdapter.TAB_REWARDS -> "Rewards"
                StakingTabsAdapter.TAB_STAKE -> "Stake"
                StakingTabsAdapter.TAB_UNSTAKE -> "Withdraw"
                StakingTabsAdapter.TAB_REDELEGATE -> "Redelegate"
                StakingTabsAdapter.TAB_UNBONDING -> "Unbonding"
                else -> "Tab $position"
            }
        }.attach()

        viewPager.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                super.onPageSelected(position)
                // Show the info panel only on the Rewards tab; hide it on Stake/
                // Withdraw/Unbonding so the amount input isn't pushed under the keyboard.
                if (position == StakingTabsAdapter.TAB_REWARDS) {
                    infoSection?.visibility = View.VISIBLE
                    refreshStakingData() // pull fresh staked/APR/earnings each time it's shown
                } else {
                    infoSection?.visibility = View.GONE
                }
            }
        })
    }

    fun refreshStakingData() {
        if (!isAdded || context == null) return
        lifecycleScope.launch {
            try {
                val address = SecureWalletManager.getWalletAddress(requireContext())
                val (staked, total) = withContext(Dispatchers.IO) {
                    val stakedBase = if (address.isNullOrEmpty()) BigInteger.ZERO else
                        Staking.delegations(address).fold(BigInteger.ZERO) { acc, d ->
                            acc + (d.amount.toBigIntegerOrNull() ?: BigInteger.ZERO)
                        }
                    val totalBase = Staking.totalBonded().toBigIntegerOrNull() ?: BigInteger.ZERO
                    stakedBase to totalBase
                }
                stakedBalance = staked.toDouble() / 1_000_000.0
                totalStakedBalance = total.toDouble() / 1_000_000.0
                apr = if (totalStakedBalance > 0) (ANNUAL_INFLATION_ERTH / totalStakedBalance) * 100 else 0.0

                try {
                    erthPrice = ErthPriceService.fetchErthPrice()
                } catch (e: Exception) {
                    Log.e(TAG, "Error fetching ERTH price", e)
                }
                updateUI()
            } catch (e: kotlinx.coroutines.CancellationException) {
                throw e
            } catch (e: Exception) {
                Log.e(TAG, "Error refreshing staking data", e)
            }
        }
    }

    private fun updateUI() {
        aprText?.text = String.format("%.2f%%", apr)
        stakedAmountText?.text = String.format("%,.2f ERTH", stakedBalance)
        stakedAmountUsd?.text = erthPrice?.let { ErthPriceService.formatUSD(stakedBalance * it) } ?: ""

        totalStakedText?.text = String.format("%,.0f ERTH", totalStakedBalance)
        totalStakedUsd?.text = erthPrice?.let { ErthPriceService.formatUSD(totalStakedBalance * it) } ?: ""

        val poolShare = if (totalStakedBalance > 0) (stakedBalance / totalStakedBalance) * 100 else 0.0
        poolShareText?.text = String.format("%.4f%%", poolShare)

        val dailyEarnings = (stakedBalance * apr / 100) / 365
        dailyEarningsText?.text = String.format("%.2f ERTH", dailyEarnings)
        dailyEarningsUsd?.text = erthPrice?.let { ErthPriceService.formatUSD(dailyEarnings * it) } ?: ""
    }

    override fun onResume() {
        super.onResume()
        refreshStakingData()
    }

    override fun onDestroy() {
        super.onDestroy()
        if (transactionSuccessReceiver != null && context != null) {
            try {
                requireActivity().applicationContext.unregisterReceiver(transactionSuccessReceiver)
            } catch (e: Exception) { }
        }
    }
}
